File: src/main/golang/leetcode/editor/cn/[95]Unique Binary Search Trees II.go
package lc

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func generateTrees(n int) []*TreeNode {

	var build func(l, r int) []*TreeNode
	build = func(l, r int) []*TreeNode {
		if l > r {
			return []*TreeNode{nil}
		}
		var ans []*TreeNode
		for i := l; i <= r; i++ {
			leftChilds := build(l, i-1)
			rightChilds := build(i+1, r)
			for _, left := range leftChilds {
				for _, right := range rightChilds {
					node := &TreeNode{Val: i}
					node.Left = left
					node.Right = right
					ans = append(ans, node)
				}
			}
		}
		return ans
	}
	return build(1, n)
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[95]Unique Binary Search Trees II.py
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        def generate(start, end):
            if start > end:
                return [None]
            all_trees = []
            for i in range(start, end + 1):
                left_trees = generate(start, i - 1)
                right_trees = generate(i + 1, end)
                for l in left_trees:
                    for r in right_trees:
                        cur = TreeNode(i)
                        cur.left = l
                        cur.right = r
                        all_trees.append(cur)
            return all_trees
        return generate(1, n) if n else []


# leetcode submit region end(Prohibit modification and deletion)
if __name__ == '__main__':
    sol = Solution()
    sol.generateTrees(3)
